sw/host/opentitantool/src/command/console.rs

```diff
@@ -0,0 +1,237 @@
+    logfile: Option<String>,
+
+    #[structopt(short, long, help = "Exit after a timeout in seconds.")]
+    timeout: Option<u64>,
+
+    #[structopt(long, help = "Exit with success if the specified regex is matched.")]
+    exit_success: Option<String>,
+
+    #[structopt(long, help = "Exit with failure if the specified regex is matched.")]
+    exit_failure: Option<String>,
+}
+
+impl CommandDispatch for Console {
+    fn run(&self, transport: &mut dyn Transport) -> Result<Option<Box<dyn Serialize>>> {
+        // We need the UART for the console command to operate.
+        transport.capabilities().request(Capability::UART).ok()?;
+        let mut stdout = std::io::stdout();
+        let mut stdin = std::io::stdin();
+
+        // Set up resources specified by the command line parameters.
+        let mut console = InnerConsole {
+            logfile: self.logfile.as_ref().map(File::create).transpose()?,
+            deadline: self.timeout.map(|t| Instant::now() + Duration::new(t, 0)),
+            exit_success: self
+                .exit_success
+                .as_ref()
+                .map(|s| Regex::new(s.as_str()))
+                .transpose()?,
+            exit_failure: self
+                .exit_failure
+                .as_ref()
+                .map(|s| Regex::new(s.as_str()))
+                .transpose()?,
+            ..Default::default()
+        };
+
+        if !self.quiet {
+            println!("Starting interactive console");
+            println!("[CTRL+C] to exit.\n");
+        }
+        {
+            // Put the terminal into raw mode.  The tty guard will restore the
+            // console settings when it goes out of scope.
+            let _stdin_guard = if isatty(stdin.as_raw_fd())? {
+                let mut guard = TtyModeGuard::new(stdin.as_raw_fd())?;
+                guard.set_raw_mode()?;
+                Some(guard)
+            } else {
+                None
+            };
+            let _stdout_guard = if isatty(stdout.as_raw_fd())? {
+                let mut guard = TtyModeGuard::new(stdout.as_raw_fd())?;
+                guard.set_raw_mode()?;
+                Some(guard)
+            } else {
+                None
+            };
+            console.interact(transport, &mut stdin, &mut stdout)?;
+        }
+        if !self.quiet {
+            println!("\n\nExiting interactive console.");
+        }
+        Ok(None)
+    }
+}
+
+#[derive(Default)]
+struct InnerConsole {
+    logfile: Option<File>,
+    deadline: Option<Instant>,
+    exit_success: Option<Regex>,
+    exit_failure: Option<Regex>,
+    buffer: String,
+}
+
+enum ExitStatus {
+    None,
+    ExitSuccess,
+    ExitFailure,
+}
+
+impl InnerConsole {
+    const CTRL_C: u8 = 3;
+    const BUFFER_LEN: usize = 1024;
+
+    // Runs an interactive console until CTRL_C is received.
+    fn interact<R>(
+        &mut self,
+        transport: &mut dyn Transport,
+        stdin: &mut R,
```

Comment:
You can write `&mut impl (Read + AsRawFd)`.

Reply:
Done.